plugins/chat/app/queries/chat/thread_participant_query.rb

```diff
@@ -0,0 +1,98 @@
+  # of:
+  #
+  # - Participant 1 & 2 - The most frequent messagers in the thread.
+  # - Participant 3 - The most recent messager in the thread.
+  #
+  # This result should be cached to avoid unnecessary queries,
+  # since the participants will not often change for a thread,
+  # and if there is a delay in updating them based on message
+  # count it is not a big deal.
+  class ThreadParticipantQuery
+    # @param thread_ids [Array<Integer>] The IDs of the threads to query.
+    # @return [Hash<Integer, Hash>] A hash of thread IDs to participant data.
+    def self.call(thread_ids:)
+      # We only want enough data for BasicUserSerializer, since the participants
+      # are just showing username & avatar.
+
+      thread_messager_stats = DB.query(<<~SQL, thread_ids: thread_ids)
+        SELECT thread_messager_stats.*, users.username, users.name, users.uploaded_avatar_id FROM (
+          SELECT chat_messages.thread_id, chat_messages.user_id, COUNT(*) AS message_count,
+            ROW_NUMBER() OVER (PARTITION BY chat_messages.thread_id ORDER BY COUNT(*) DESC) AS row_number
+          FROM chat_messages
+          INNER JOIN chat_threads ON chat_threads.id = chat_messages.thread_id
+          INNER JOIN user_chat_thread_memberships ON user_chat_thread_memberships.thread_id = chat_threads.id
+            AND user_chat_thread_memberships.user_id = chat_messages.user_id
+          WHERE chat_messages.thread_id IN (:thread_ids)
+          AND chat_messages.deleted_at IS NULL
+          GROUP BY chat_messages.thread_id, chat_messages.user_id
+        ) AS thread_messager_stats
+        INNER JOIN users ON users.id = thread_messager_stats.user_id
+        ORDER BY thread_messager_stats.thread_id ASC, thread_messager_stats.message_count DESC, thread_messager_stats.user_id ASC
+      SQL
+
+      most_recent_messagers = DB.query(<<~SQL, thread_ids: thread_ids)
+        SELECT DISTINCT ON (thread_id) chat_messages.thread_id, chat_messages.user_id,
+          users.username, users.name, users.uploaded_avatar_id
+        FROM chat_messages
+        INNER JOIN chat_threads ON chat_threads.id = chat_messages.thread_id
+        INNER JOIN user_chat_thread_memberships ON user_chat_thread_memberships.thread_id = chat_threads.id
+          AND user_chat_thread_memberships.user_id = chat_messages.user_id
+        INNER JOIN users ON users.id = chat_messages.user_id
+        WHERE chat_messages.thread_id IN (:thread_ids)
+        AND chat_messages.deleted_at IS NULL
+        ORDER BY chat_messages.thread_id ASC, chat_messages.created_at DESC
+      SQL
+      most_recent_messagers =
+        most_recent_messagers
+          .map do |mrm|
+            [
+              mrm.thread_id,
+              {
+                id: mrm.user_id,
+                username: mrm.username,
+                name: mrm.name,
+                uploaded_avatar_id: mrm.uploaded_avatar_id,
+              },
+            ]
+          end
+          .to_h
```

Comment:
```suggestion
        most_recent_messagers
          .reduce({}) do |acc, mrm|
            acc[mrm.thread_id] = {
              id: mrm.user_id,
              username: mrm.username,
              name: mrm.name,
              uploaded_avatar_id: mrm.uploaded_avatar_id,
            }

            acc
          end
```

Reply:
Oh nice ty, I always forget about this way to build up a hash :+1: